src/onboarding/map.ts

```diff
@@ -0,0 +1,163 @@
+import { initializeAccount } from 'src/account/actions'
+import {
+  choseToRestoreAccountSelector,
+  recoveringFromStoreWipeSelector,
+} from 'src/account/selectors'
+import {
+  numberVerifiedCentrallySelector,
+  skipVerificationSelector,
+  supportedBiometryTypeSelector,
+} from 'src/app/selectors'
+import { setHasSeenVerificationNux } from 'src/identity/actions'
+import * as NavigationService from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { store } from 'src/redux/store'
+
+interface GetOnboardingMapProps {
+  navigate: typeof NavigationService.navigate | ((screen: Screens) => void)
+  popToScreen: typeof NavigationService.popToScreen | ((screen: Screens) => void)
+  navigateHome: typeof NavigationService.navigateHome | (() => void)
+  navigateClearingStack:
+    | typeof NavigationService.navigateClearingStack
+    | ((screen: Screens) => void)
+  dispatch: (action: any) => void
+}
+
+/**
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getStepValues(screen: Screens) {
+  let stepCounter = 0
+  let totalCounter = 0
+  let reachedScreeen = false
+  let currentScreen: Screens = Screens.NameAndPicture
+
+  const navigate = (s: Screens) => {
+    totalCounter++
+    currentScreen = s
+    if (!reachedScreeen) {
+      stepCounter++
+    }
+    if (currentScreen === screen) {
+      reachedScreeen = true
+    }
+  }
+
+  const navigateHome = () => {
+    currentScreen = Screens.WalletHome
+  }
+
+  const onBoardingMap = _getOnboardingMap({
+    navigate,
+    popToScreen: () => {
+      // no-op
+    },
+    navigateHome,
+    navigateClearingStack: navigate,
+    dispatch: () => {
+      // no-op
+    },
+  })
+
+  while (currentScreen !== Screens.WalletHome) {
+    onBoardingMap[currentScreen]?.next()
+  }
+  return {
+    totalCounter,
+    stepCounter,
+  }
+}
+
+export function navigateNextOnboarding(screen: Screens) {
+  const onBoardingMap = _getOnboardingMap({
+    navigate: NavigationService.navigate,
+    popToScreen: NavigationService.popToScreen,
+    navigateHome: NavigationService.navigateHome,
+    navigateClearingStack: NavigationService.navigateClearingStack,
+    dispatch: store.dispatch,
+  })
+  onBoardingMap[screen]?.next()
+}
+
+function _getOnboardingMap({
```

Comment:
this is really interesting! I love how it uses dependency injection to make it possible to just traverse the graph without creating a bunch of side effects.

I almost want to take it a step further where it takes all this state as a parameter-- or maybe uses some TS magic to only require the state that's needed from the start screen you're at??-- and returns a screen you need to go to, plus some way of telling what side effects need to take place if you're going to that screen in a "real" onboarding navigation flow. it could be just a function like "doSideEffects". the `navigateClearingStack` use case is annoying, but maybe there's some way of taking what that does and separating out the effectful part into the `doSideEffects` piece.

It's already in pretty good shape so it's hard to say what's an improvement about my idea exactly. Taking state as a parameter makes it a pure function, so that's good. I guess separating the side effects out into a function on the return value makes it feel more fair to use it for `getStepValues`. 

I think my main feeling is that this isn't really a "map"; it's more like a state-dependent set of edges between screens that have a start screen, an end screen, and a side effect for the transition between the screens (which you may or may not want to use!).